### ISSUE TITLE
Double data splitting in get_train_dataloader for multi-gpu setting?

### ISSUE THREAD

**Issue by tnnandi** [huggingface/transformers#34365]
### System Info

- `transformers` version: 4.45.2
- Platform: Linux-5.14.21-150500.55.49-default-x86_64-with-glibc2.31
- Python version: 3.10.15
- Huggingface_hub version: 0.26.1
- Safetensors version: 0.4.5
- Accelerate version: 1.0.1
- Accelerate config:    - compute_environment: LOCAL_MACHINE
        - distributed_type: DEEPSPEED
        - use_cpu: False
        - debug: False
        - num_processes: 8
        - machine_rank: 0
        - num_machines: 2
        - main_process_ip: 123.12.34.45
        - main_process_port: 2345
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: False
        - deepspeed_config: {'deepspeed_config_file': '/grand/GeomicVar/tarak/Geneformer_oct_2024/Geneformer/run/ds_config_zero.json', 'deepspeed_multinode_launcher': 'standard', 'zero3_init_flag': False}
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
- PyTorch version (GPU?): 2.5.0+cu124 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: NVIDIA A100-SXM4-40GB

### Who can help?

@muellerz @SunMarc 

In get_train_dataloader(), for a sample size of N with M GPUs, len(DataLoader(train_dataset, **dataloader_params)) is N/M (As expected), but len(self.accelerator.prepare(DataLoader(train_dataset, **dataloader_params))) is N/(MxM).  Since a sampler is being used, the dataset should be split into the M GPUs appropriately, and I'm not being able to understand if accelerate.prepare() should be carrying out another splitting step.  When printing the batches (using a batch size of 1 per GPU), I can only view N/(M*M) samples per batch, and the rest of the samples [N/M - N/(MxM)] are not part of any batch. 

```
        dataloader_params = {
            "batch_size": self._train_batch_size,
            "collate_fn": data_collator,
            "num_workers": self.args.dataloader_num_workers,
            "pin_memory": self.args.dataloader_pin_memory,
            "persistent_workers": self.args.dataloader_persistent_workers,
        }

        if not isinstance(train_dataset, torch.utils.data.IterableDataset):
            dataloader_params["sampler"] = self._get_train_sampler()
            dataloader_params["drop_last"] = self.args.dataloader_drop_last
            dataloader_params["worker_init_fn"] = seed_worker
            dataloader_params["prefetch_factor"] = self.args.dataloader_prefetch_factor

        return self.accelerator.prepare(DataLoader(train_dataset, **dataloader_params))
```

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```
        dataloader_params = {
            "batch_size": self._train_batch_size,
            "collate_fn": data_collator,
            "num_workers": self.args.dataloader_num_workers,
            "pin_memory": self.args.dataloader_pin_memory,
            "persistent_workers": self.args.dataloader_persistent_workers,
        }

        if not isinstance(train_dataset, torch.utils.data.IterableDataset):
            dataloader_params["sampler"] = self._get_train_sampler()
            dataloader_params["drop_last"] = self.args.dataloader_drop_last
            dataloader_params["worker_init_fn"] = seed_worker
            dataloader_params["prefetch_factor"] = self.args.dataloader_prefetch_factor
       
       ###########  print statements to confirm dataloader sizes ##############
       print(len(DataLoader(train_dataset, **dataloader_params))
       print(self.accelerator.prepare(DataLoader(train_dataset, **dataloader_params))) 

        return self.accelerator.prepare(DataLoader(train_dataset, **dataloader_params))
```

### Expected behavior

If the N samples are being distributed appropriately to M GPUs, the above function should return a dataloader with N/M samples. Instead it returns a dataloader of length N/(M*M)

**Comment by SunMarc** [comment]
Hey @tnnandi, thanks for the report. Could you share a minimal reproducer with logs that shows this behavior. This will be very helpful when debugging ! cc @muellerzr

**Comment by tnnandi** [comment]
Thanks! Will do that shortly.

**Comment by tnnandi** [comment]
**Command to run**: accelerate launch --config_file ./accelerate_config_singlenode.without_ds.yaml test_samples_per_batch.py

In the script below, I have tried to replicate the issue by creating a custom trainer class that closely follows the HF Trainer class in context of the issue I raised (i.e., includes the relevant functions to reproduce the issue).

A dataset with 40 samples is created below, that when using 4 GPUs should be split into 10 samples on each GPU. While that is true when get_train_dataloader() in the CustomPretrainer class returns _DataLoader(self.train_dataset, **dataloader_params)_ , only 3 samples are processed per GPU when the same function returns _self.accelerator.prepare(DataLoader(self.train_dataset, **dataloader_params))_[you can confirm from len(train_dataloader) or the prints from the training loop]

Code for _test_samples_per_batch.py_:

`import os
import random
import numpy as np
import torch
from torch.utils.data import DataLoader, Dataset, DistributedSampler
from transformers import LlamaForCausalLM, LlamaConfig
from accelerate import Accelerator
from accelerate.logging import get_logger
from dataclasses import dataclass

logger = get_logger(__name__)


@dataclass
class TrainingConfig:
    output_dir: str = './results'
    per_device_train_batch_size: int = 1
    gradient_accumulation_steps: int = 1
    num_train_epochs: int = 1
    dataloader_num_workers: int = 1
    dataloader_pin_memory: bool = True
    dataloader_persistent_workers: bool = False
    dataloader_drop_last: bool = False


def setup_environment(base_dir='./test_dataloader', seed_val=42):
    try:
        os.makedirs(base_dir, exist_ok=True)
    except Exception as e:
        print(f"Error creating directory: {e}")

    random.seed(seed_val)
    np.random.seed(seed_val)
    torch.manual_seed(seed_val)
    torch.cuda.manual_seed_all(seed_val)


class RandomDataset(Dataset):
    def __init__(self, num_samples, seq_length, vocab_size):
        self.num_samples = num_samples
        self.seq_length = seq_length
        self.vocab_size = vocab_size

    def __len__(self):
        return self.num_samples

    def __getitem__(self, idx):
        input_ids = torch.randint(0, self.vocab_size, (self.seq_length,))
        return {
            'input_ids': input_ids,
            'attention_mask': torch.ones(self.seq_length),
            'labels': input_ids.clone()
        }


class CustomPretrainer:
    def __init__(self, train_dataset, accelerator, batch_size, training_config):
        self.train_dataset = train_dataset
        self.accelerator = accelerator
        self._train_batch_size = batch_size
        self.config = training_config

    def _get_train_sampler(self):
        if self.accelerator.num_processes > 1:
            return DistributedSampler(
                self.train_dataset,
                num_replicas=self.accelerator.num_processes,
                rank=self.accelerator.process_index,
                shuffle=True
            )
        return None

    def get_train_dataloader(self):
        if self.train_dataset is None:
            raise ValueError("Trainer: training requires a train_dataset.")

        sampler = self._get_train_sampler()
        dataloader_params = {
            "batch_size": self._train_batch_size,
            "num_workers": self.config.dataloader_num_workers,
            "pin_memory": self.config.dataloader_pin_memory,
            "persistent_workers": self.config.dataloader_persistent_workers,
            "drop_last": self.config.dataloader_drop_last,
            "sampler": sampler
        }

        # return DataLoader(self.train_dataset, **dataloader_params)
        return self.accelerator.prepare(DataLoader(self.train_dataset, **dataloader_params))


def main():
    accelerator = Accelerator()
    print("accelerator.num_processes at start:", accelerator.num_processes)

    # Setup environment
    setup_environment()

    # Model configuration
    config = LlamaConfig(
        hidden_size=128,
        num_hidden_layers=32,
        intermediate_size=256,
        max_position_embeddings=2 ** 12,
        model_type="llama",
        bos_token_id=2,
        eos_token_id=3,
        pad_token_id=0,
        num_attention_heads=8,
        vocab_size=50000,
    )

    # Create dataset
    train_dataset = RandomDataset(
        num_samples=40,
        seq_length=5,
        vocab_size=config.vocab_size
    )

    # Create training config
    training_config = TrainingConfig(
        per_device_train_batch_size=1,
        gradient_accumulation_steps=1
    )

    print("accelerator.num_processes after config:", accelerator.num_processes)

    model = LlamaForCausalLM(config)
    optimizer = torch.optim.AdamW(model.parameters(), lr=0.001)

    # Prepare model and optimizer
    model, optimizer = accelerator.prepare(model, optimizer)

    # Instantiate trainer and create DataLoader
    trainer = CustomPretrainer(
        train_dataset=train_dataset,
        accelerator=accelerator,
        batch_size=training_config.per_device_train_batch_size,
        training_config=training_config
    )

    train_dataloader = trainer.get_train_dataloader()

    print("len(train_dataloader): ", len(train_dataloader))

    # Training loop
    device = accelerator.device
    print("Device: ", device)
    for epoch in range(training_config.num_train_epochs):
        # Set epoch to ensure proper shuffling with DistributedSampler
        print(train_dataloader.sampler)
        if isinstance(train_dataloader.sampler, DistributedSampler):
            train_dataloader.sampler.set_epoch(epoch)

        epoch_loss = 0.0
        num_steps = 0

        for step, batch in enumerate(train_dataloader):
            batch = {k: v.to(device) for k, v in batch.items()}
            print(f"{batch['input_ids'].shape} from device {accelerator.process_index}")
            # print(f"{batch['input_ids']} from device {accelerator.process_index}")
            print("model.device: ", model.device)
            outputs = model(**batch)
            loss = outputs.loss

            accelerator.backward(loss)
            optimizer.step()
            optimizer.zero_grad()

            epoch_loss += loss.item()
            num_steps += 1
            print(f"Epoch: {epoch + 1}, Step: {step + 1} of {len(train_dataloader)}, Process Rank: {accelerator.process_index}")

            del batch, outputs, loss
            torch.cuda.empty_cache()

        print(f"Epoch {epoch + 1} completed with average loss: {epoch_loss / num_steps:.4f}")

    print("Training completed.")


if __name__ == "__main__":
    main()
`

Content of accelerate config file _accelerate_config_singlenode.without_ds.yaml_:
```
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: MULTI_GPU
downcast_bf16: 'no'
enable_cpu_affinity: false
machine_rank: 0
main_training_function: main
num_machines: 1
num_processes: 4
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

**Comment by muellerzr** [comment]
Thanks, I have some time now so let me look into this over the next few hours and get back with you

**Comment by muellerzr** [comment]
actually I see right away. You're using the `DistributedSampler`. You should **not** be doing this because Accelerate's dataloaders do this. Let accelerate handle all of it that you pass to `prepare`.

**Comment by tnnandi** [comment]
@muellerzr Thank you very much for pointing it out! I really appreciate the work you and the HF team is doing.

**Comment by Rocketknight1** [comment]
Closing this for now, feel free to reopen it if you have any other problems!
